refs #177 added rake task to set slug for legacy data

anuja-joshi opened the pull request:
- Added rake task `rake mongoid_slug:set` that will set slug for existing records, if not already set.
- We can pass optional parameters to task like `rake mongoid_slug:set[Model1,Model2]` so that slug will get set for records of only these models.
- To load this task into `Rails`, I loaded task using `Rails::Railtie` so this task will be directly available to `Rails` projects. For other frameworks user have to load tasks file into their rake stack.
- Added associated test cases.

dblock commented on README.md:
````diff
@@ -72,6 +72,16 @@ post = Post.find '50b1386a0482939864000001' # Finds by bson ids
 ```
 [Examine slug.rb](lib/mongoid/slug.rb) for all available options.
 
+To set slug for existing records run following rake task :
````
Extra space before `:`.

dblock commented on README.md:
````diff
@@ -72,6 +72,16 @@ post = Post.find '50b1386a0482939864000001' # Finds by bson ids
+```ruby
+rake mongoid_slug:set
+```
+You can pass model names as an option for which you want to set slug:
````
probably to set slugs (plural)?

dblock commented:
Why does this have to be Rais-specific? The only rails-specific thing I see here is the Railtie, which is really just a convenience to load Rake tasks. No?

anuja-joshi commented:
@dblock Yes, task is not Rails specific. I added `Railtie` just to reduce efforts of explicitly loading this rake task file into Rails project's stack.

I corrected README.

dblock commented on lib/tasks/mongoid_slug.rake:
```diff
@@ -0,0 +1,16 @@
+    unless Mongoid::Slug.classes.blank?
+      models  = args.extras
+      klasses = Mongoid::Slug.classes
+      klasses = (klasses.map(&:to_s) & models.map(&:classify)).map(&:constantize) unless models.blank?
```
`unless models.blank?` should really be `if models.any?`, but it's NBD

dblock commented on spec/tasks/mongoid_slug_rake_spec.rb:
```diff
@@ -0,0 +1,52 @@
+  context 'when models parameter is not passed' do
+    let!(:uninitalized_slug3) { UninitalizedSlugThird.create(name: 'uninitalized_slug3') }
+
+    it 'Goes though all documents and sets slug if not already set' do
```
we lowercase these, also nbd

dblock commented on spec/tasks/mongoid_slug_rake_spec.rb:
```diff
@@ -0,0 +1,52 @@
+    let!(:uninitalized_slug2) { UninitalizedSlugSecond.create(name: 'uninitalized_slug2') }
+
+    it 'Goes though all documents of passed models and sets slug if not already set' do
+      class UninitalizedSlugFirst
```
You have these redefined here, they are already in `spec/models`. 

If you define them inline (which could be fine, too), use:

```ruby
let(:uninitialized_slug_klass) { Class.new do
    include Mongoid::Slug
    slug :name
end
}
```

Might need a `store_in` or something like that since the class doesn't have a name.

anuja-joshi replied:
@dblock I didn't redefine, I reopened these classes to include  `Mongoid::Slug`, as I wanted to simulate situation where few objects of class are already created and `Mongoid::Slug` is included later. In such case slug do not initialize for previously created objects and running task will set slugs for such objects.
Correct me, if this is not proper way.

dblock replied:
I see. Yes, because spec files are loaded in any order and executed in some other you can get in a situation where a subsequent spec depends on the original behavior of the class, and since you reopened and modified it, you now have an unexpected result. Right now this is not a problem in the spec, but in two months someone will update these and we won't know where to begin debugging it ;)

Rewrite these with dynamic classes, ie. `Class.new do ...`?

dblock commented:
Would you please correct the minor stuff and the double-definition. This will be good to merge. Thx.

dblock commented:
This also needs an entry in CHANGELOG, please.

dblock commented on spec/tasks/mongoid_slug_rake_spec.rb:
```diff
@@ -0,0 +1,52 @@
+    let!(:uninitalized_slug1) { UninitalizedSlugFirst.create(name: 'uninitalized_slug1') }
+    let!(:uninitalized_slug2) { UninitalizedSlugSecond.create(name: 'uninitalized_slug2') }
+
+    it 'Goes though all documents of passed models and sets slug if not already set' do
```
Goes -> goes

dblock commented:
I think you also want a `load_tasks` method like [here](https://github.com/rosenfeld/active_record_migrations/blob/fc7a8e4304417fdff3ee5ff24a5385e73cfdb7ab/lib/active_record_migrations.rb) so it can be used outside of Rails. And document it please.

dblock commented on lib/tasks/mongoid_slug.rake:
```diff
@@ -0,0 +1,16 @@
+namespace :mongoid_slug do
+  desc 'Goes though all documents and sets slug if not already set'
+  task set: :environment do |_, args|
+    ::Rails.application.eager_load! if defined?(Rails)
```
I think this should go into the Railtie withotu the `if`.

anuja-joshi replied:
@dblock if we add `::Rails.application.eager_load!` in Railtie then it will get execute unnecessarily whenever Rails console or server is started so I kept it here in rake task.

dblock replied:
I see.

dblock commented:
Btw, there's a way to find all classes that mix in `Mongoid::Slug` without storing it. Maybe better? Not sure.

https://www.ruby-forum.com/topic/163430

```ruby
Module.constants.find_all{ |k| k != k.upcase ? Mongoid::Slug > (Object.const_get k) : nil}
```

dblock commented:
Also this motivated me to write https://github.com/dblock/module-mixins if you want to try that :)

anuja-joshi commented:
@dblock yes, approach suggested by you for finding all classes that includes `Mongoid::Slug` is better than storing them in variable, will rectify PR.

dblock commented on CHANGELOG.md:
```diff
@@ -1,8 +1,8 @@
 # CHANGELOG
 
-## Next
+## 5.1.2
```
Put back Next please.

dblock commented on CHANGELOG.md:
```diff
@@ -1,8 +1,8 @@
-## Next
+## 5.1.2
 
-* Your contribution here.
```
Put back your contribution here please.

dblock commented on lib/tasks/mongoid_slug.rake:
```diff
@@ -0,0 +1,19 @@
+      klasses = (klasses.map(&:to_s) & models.map(&:classify)).map(&:constantize) if models.any?
+      klasses.each do |klass|
+        # set slug for objects having blank slug
+        klass.each { |object| object.set_slug! unless object.slugs? }
```
This is a total nitpick feel free not to change, unless maybe this is more readable:

```ruby
klass.each do |object|
   next if object.slugs?
   object.set_slug!
end
```

dblock commented on lib/tasks/mongoid_slug.rake:
```diff
@@ -0,0 +1,19 @@
+        klass.each { |object| object.set_slug! unless object.slugs? }
+      end
+    end
+    p 'Completed mongoid_slug:set'
```
Remove this, don't rely on STDOUT, and we don't have a logger.

dblock commented on spec/tasks/mongoid_slug_rake_spec.rb:
```diff
@@ -0,0 +1,73 @@
+
+  context 'when models parameter is passed' do
+    before :all do
+      UninitalizedSlugFirst = Class.new do
```
You can do this via a `let` too, probably cleaner, not important.

dblock commented:
I have a few nitpicks, and also please check the build.

dblock commented:
Bump. Squash/rebase this and fix the build, I'm happy to merge.

anuja-joshi commented:
@dblock rebased with master and squashed commits.

dblock commented:
Merged via https://github.com/digitalplaywright/mongoid-slug/commit/8abb60d6c263f532e51b1ea064b88e6df25b2658, thanks! Just amended the CHANGELOG slightly. 

I think we need a `Mongoid::Slug.load_tasks` for non-Rails environments, and you would also replace `load File.expand_path('../../../lib/tasks/mongoid_slug.rake', __FILE__)` in the tests with that. Want to try to code that up?
